src/docbuilder/metadata/lib.rs

```diff
@@ -0,0 +1,631 @@
+    /// If you just want the default settings, use [`Metadata::default()`][Default::default].
+    ///
+    /// [`from_str`]: std::str::FromStr
+    pub fn from_manifest<P: AsRef<Path>>(path: P) -> Result<Metadata, MetadataError> {
+        use std::{str::FromStr, fs};
+        let buf = fs::read_to_string(path)?;
+        Metadata::from_str(&buf).map_err(Into::into)
+    }
+
+    /// Return the targets that should be built.
+    ///
+    /// The `default_target` will never be one of the `other_targets`.
+    pub fn targets(&self) -> BuildTargets<'_> {
+        let default_target = self
+            .default_target
+            .as_deref()
+            // Use the first element of `targets` if `default_target` is unset and `targets` is non-empty
+            .or_else(|| {
+                self.targets
+                    .as_ref()
+                    .and_then(|targets| targets.iter().next().map(String::as_str))
+            })
+            .unwrap_or(HOST_TARGET);
+
+        // Let people opt-in to only having specific targets
+        let mut targets: HashSet<_> = self
+            .targets
+            .as_ref()
+            .map(|targets| targets.iter().map(String::as_str).collect())
+            .unwrap_or_else(|| DEFAULT_TARGETS.iter().copied().collect());
+
+        targets.remove(&default_target);
+        BuildTargets {
+            default_target,
+            other_targets: targets,
+        }
+    }
+
+    /// Return the arguments that should be passed to `cargo`.
+    ///
+    // TODO: maybe it shouldn't?
+    /// This will always include `doc --lib --no-deps`.
+    /// This will never include `--target`.
+    ///
+    /// Note that this does not necessarily reproduce the HTML _output_ of docs.rs exactly.
+    /// For example, the links may point somewhere different than they would on docs.rs.
+    /// However, rustdoc will see exactly the same code as it would on docs.rs, even counting `cfg`s.
+    pub fn cargo_args(&self) -> Vec<String> {
+        let mut cargo_args: Vec<String> = vec!["doc".into(), "--lib".into(), "--no-deps".into()];
+
+        if let Some(features) = &self.features {
+            cargo_args.push("--features".into());
+            cargo_args.push(features.join(" "));
+        }
+
+        if self.all_features {
+            cargo_args.push("--all-features".into());
+        }
+
+        if self.no_default_features {
+            cargo_args.push("--no-default-features".into());
+        }
+
+        cargo_args
+    }
+
+    /// Return the environment variables that should be set when building this crate.
+    pub fn environment_variables(&self) -> HashMap<&'static str, String> {
+        let joined = |v: &Option<Vec<_>>| v.as_ref()
+            .map(|args| args.join(" "))
+            .unwrap_or_default();
+
+        let mut map = HashMap::new();
+        map.insert("RUSTFLAGS", joined(&self.rustc_args));
+        map.insert("RUSTDOCFLAGS", joined(&self.rustdoc_args));
```

Comment:
We're not adding `-Z unstable-options`, so the user's `RUSTDOCFLAGS` could not work. Also, we're missing a lot of other flags set by docs.rs.

Reply:
The other flags are set in the rustwide builder: https://github.com/rust-lang/docs.rs/pull/1000/files#diff-8fac5c4dd8b0132a470c525431c6c4bcR609-R611. My reasoning is that crater doesn't need these flags, so they don't need to be part of the library (and moving over anything besides `-Z unstable-options` will be annoying because it needs access to the full TOML, not just `[package.metadata.docs.rs]`). Do you want -Z unstable-options in the library?

Reply:
So, this is tricky: ideally I'd prefer to have all the flags being returned by `Metadata`, so Crater and other tools can fully test the exact setup docs.rs uses. On the other hand, docs.rs relies on unstable flags, and we often run Crater on the beta branch, so adding `-Z unstable-options` would prevent this from being used in beta runs :/

Reply:
If it helps, the only difference this should make is for users passing unstable `RUSTDOCFLAGS` in CI, which seems like a pretty niche thing to do. Most of the unstable options affect only what documentation is _generated_, not the code rustdoc sees.

Reply:
Most of the rustdocflags I've seen in the wild enable unstable compiler features via `cfg_attr` (mostly `#![feature(doc_cfg)]`), so do affect the code rustdoc sees.

Reply:
Or I guess we could add everything to `.environment_variables()` and trick Crater to think it's on nightly even on beta when the "docs.rs mode" is used.

Reply:
`stable()` and `unstable()` doesn't seem great because we don't know which of the user flags are which, we'd have to put them all in `unstable()`. I think the best way is to use `RUSTC_BOOTSTRAP` in crater, yeah.

Reply:
Ok, let's move all environment variables handling in the library then.

Reply:
Sorry - I was suggesting to move `-Z unstable-options` into the library, not `--extern-html-root`. Aside from the mess that makes on the command line, it's not relevant at all to crater since it will still see the same code either way, only the generated links will be different. And it will be very hard to make that part of the library.

Reply:
Pushed a commit adding `-Z unstable-options`, let me know if you still have concerns. IMO the remaining options don't make sense to use with crater:
- `-j2`
- `--target`
- `--static-root-path /`
- `--cap-lints warn`
- `--resource-suffix`
- `--extern-html-root-url`